x/mint/types/minter.go

```diff
@@ -92,7 +93,14 @@ func (m Minter) NextAnnualProvisions(_ Params, totalSupply sdk.Int) sdk.Dec {
 
 // BlockProvision returns the provisions for a block based on the annual
 // provisions rate.
-func (m Minter) BlockProvision(params Params) sdk.Coin {
+func (m Minter) BlockProvision(params Params, totalSupply sdk.Int) sdk.Coin {
```

Comment:
What is `totalSupply` here?

Reply:
So, block rewards must be integer. 

With some chain configuration it can happen (and actually is very common) that annual_provision is not a multiple of blocks per year, resulting in a single block reward that would be N ujuno and something. Because 1 ujuno is the minimum amount we can always have, we had to round the result of that division (line 105). 

This leads to phases that can have 1 block more than what it should, by knowing the total supply we can calculate the difference and put it back on track. 

If we don't do that, on mainnet will not be a hug problem, and the maximum supply after 12 years would be max 48 JUNO (4 Juno * phase count) off the target. On testnets or chains where blocks per year is much smaller, therefore single block rewards bigger, the difference is way more noticeable

I can explain better with some example if it's not clear

Reply:
Right, that makes sense. Also means that although that parameter is there, it's in most cases going to have _no effect_ on the outcome of the function, right?

Reply:
that's right! most of the times that check against target supply of the phase is false, happens just once per phase

Reply:
I added a very arbitrary test to show what the effect of that would be in the event it was hit 🤷